src/wallet/wallet_rpc_server.cpp

```diff
@@ -337,30 +337,51 @@ namespace tools
     if (!m_wallet) return not_open(er);
     try
     {
-      res.balance = m_wallet->balance(req.account_index);
-      res.unlocked_balance = m_wallet->unlocked_balance(req.account_index);
-      res.multisig_import_needed = m_wallet->multisig() && m_wallet->has_multisig_partial_key_images();
-      std::map<uint32_t, uint64_t> balance_per_subaddress = m_wallet->balance_per_subaddress(req.account_index);
-      std::map<uint32_t, uint64_t> unlocked_balance_per_subaddress = m_wallet->unlocked_balance_per_subaddress(req.account_index);
-      std::vector<tools::wallet2::transfer_details> transfers;
-      m_wallet->get_transfers(transfers);
-      std::set<uint32_t> address_indices = req.address_indices;
-      if (address_indices.empty())
-      {
-        for (const auto& i : balance_per_subaddress)
-          address_indices.insert(i.first);
-      }
-      for (uint32_t i : address_indices)
-      {
-        wallet_rpc::COMMAND_RPC_GET_BALANCE::per_subaddress_info info;
-        info.address_index = i;
-        cryptonote::subaddress_index index = {req.account_index, info.address_index};
-        info.address = m_wallet->get_subaddress_as_str(index);
-        info.balance = balance_per_subaddress[i];
-        info.unlocked_balance = unlocked_balance_per_subaddress[i];
-        info.label = m_wallet->get_subaddress_label(index);
-        info.num_unspent_outputs = std::count_if(transfers.begin(), transfers.end(), [&](const tools::wallet2::transfer_details& td) { return !td.m_spent && td.m_subaddr_index == index; });
-        res.per_subaddress.push_back(info);
+      uint32_t account_index = req.account_index;
+      // helper loop counter variable
+      uint32_t _index = account_index;
```

Comment:
This is still ugly naming. And an unnecessary comment if sensibly renamed/refactored. Why not use variables named something like start_index, last_index? You would then set both start_index and last_index to req.account_index and then later when you check if all_accounts is requested, you can change the start_index to 0 and last_index to the last account index, incrementing start_index in your loop. Something like that would be cleaner.

Reply:
Moreover, you can get the client to get into an infinite loop by setting req.account_index to 0xffffffff.

Reply:
Indeed. The more I think about it, start and last are probably more useful too, as that would allow clients to page/batch requests.